### PR TITLE
FAD-5629 Suppressions - Upload List

### PR DESCRIPTION
**Pull request by bkemper**
_Refer to [FAD-5629](https://jira.int.messagesystems.com/browse/FAD-5629) for AC._  Note: entering a single entry is not included and will be addressed in a later ticket.

**Before**

screen shot 2018-01-25 at 3 58 45 pm

**After**

screen shot 2018-02-07 at 1 34 27 am

**How to test?**
- Use these [examples.zip](https://github.com/SparkPost/2web2ui/files/1702002/Archive.zip) files to test different states.
- Quick sanity check that recipient list upload still works as expected.

### CODE REVIEW

File: src/pages/suppressions/CreatePage.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+import { Page } from '@sparkpost/matchbox';
+
+export class CreatePage extends Component {
+  render() {
+    return (
+      <Page title="Create Suppression List">
+        Hi
```
Comment (rajumsys) [inline src/pages/suppressions/CreatePage.js]: hello :)

**Comment by bkemper** [comment]
Rebased and ready for review.

**Comment by ewandennis** [comment]
Visual review first:
 - Uploading `example-wrong-type.csv` [threw on an unhandled rejection](https://sentry.io/sparkpost/2web2ui-dev/issues/452242423/).
 - I'm not sure about the 'only for a single subaccount` checkbox.
    - Just showing the subaccount typeahead with a note saying "leave this blank to upload to the master account suppression list" might work.
    - If we continue to hide the typeahead, maybe a radio group would make both options explicit:
        - Upload to the master account
        - Upload to a subaccount
        - ^^ imperative language is also good here IMO

File: src/pages/suppressions/tests/CreatePage.test.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    },
+    submitting: false,
+    submitFailed: true
+  },
+  'renders disabled upload form when submitting': {
+    submitting: true
+  }
+});
+
+describe('CreatePage.handleSubmit', async() => {
```
Comment (ewandennis) [inline src/pages/suppressions/tests/CreatePage.test.js]: You're not testing that the correct action creators were called here. which leaves the non-subaccount case untested too. Any reason?

File: src/pages/suppressions/CreatePage.js [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+      >
+        <form onSubmit={reduxFormSubmit(this.handleSubmit)}>
+          {!submitting && submitFailed && (
+            <ApiErrorBanner
+              errorDetails={this.errorDetails}
+              message="Unable to upload your list of suppressions"
+              status="danger"
+            />
+          )}
+          <Panel title="Upload a List of Suppressions">
```
Comment (ewandennis) [inline src/pages/suppressions/CreatePage.js]: Do we need the page and panel titles? I might just pick one.

**Comment by bkemper** [comment]
Need to do some more transformations and validation before sending request to match webui.

- confirm required fields are present (recipient and type)
- typo proof the type field
- regex test recipient field
- support for deprecated fields (transactional and non_transactional ~> type, email ~> recipient)

File: src/actions/suppressions.js [hunk 4]
```diff
@@ -115,3 +116,35 @@ export function deleteSuppression(suppression) {
   });
 }
 
+// SEE: https://github.com/SparkPost/sparkpost-api-documentation/blob/master/services/suppression-list.md#insert-or-update-list-entries-put
```
[the remark this answers is not in the document]
Comment (bkemper) [reply src/actions/suppressions.js]: Addressed... 079ec707

File: src/pages/suppressions/CreatePage.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    return (
+      <Page
+        title="Add Suppressions"
+        breadcrumbAction={{ content: 'Suppressions', Component: Link, to: '/lists/suppressions' }}
+      >
+        <form onSubmit={reduxFormSubmit(this.handleSubmit)}>
+          {!submitting && submitFailed && (
+            <ApiErrorBanner
+              errorDetails={this.errorDetails}
```
[the remark this answers is not in the document]
Comment (bkemper) [reply src/pages/suppressions/CreatePage.js]: Is 50 a reasonable max?

File: src/pages/suppressions/CreatePage.js (hunk 3 above)
Comment (bkemper) [reply src/pages/suppressions/CreatePage.js]: I think we need a page title to keep a uniformed look.  How does this look without the panel title?  

screen shot 2018-02-07 at 11 48 39 am

File: src/pages/suppressions/CreatePage.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+              status="danger"
+            />
+          )}
+          <Panel title="Upload a List of Suppressions">
+            <Panel.Section>
+              <Field
+                component={FileFieldWrapper}
+                disabled={submitting}
+                name="suppressionsFile"
+                label="Comma Separated Values (CSV) File of Suppressions"
```
[the remark this answers is not in the document]
Comment (bkemper) [reply src/pages/suppressions/CreatePage.js]: Addressed... b0715041

File: src/pages/suppressions/CreatePage.js (hunk 3 above)
Comment (bkemper) [reply src/pages/suppressions/CreatePage.js]: Addressed... b0715041

File: src/helpers/validation.js [hunk 7]
```diff
@@ -34,6 +34,15 @@ export function endsWithWhitespace(value) {
   return /[\S]$/.test(value) ? undefined : 'Can\'t end in a whitespace character';
 }
 
+export function fileWithContents(file) {
```
[the remark this answers is not in the document]
Comment (bkemper) [reply src/helpers/validation.js]: Addressed... 3060911b

File: src/pages/suppressions/CreatePage.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    showAlert({ message: 'Successfully updated your suppression list', type: 'success' });
+    history.push('/lists/suppressions');
+  }
+
+  get errorDetails() {
+    const { parseError, persistError } = this.props;
+
+    return parseError
+      ? parseError.details || parseError.message
+      : _.get(persistError, 'response.data.errors[0].message', persistError.message);
```
[the remark this answers is not in the document]
Comment (bkemper) [reply src/pages/suppressions/CreatePage.js]: Addressed... 90cf2e17

File: src/pages/suppressions/tests/CreatePage.test.js (hunk 2 above)
Comment (bkemper) [reply src/pages/suppressions/tests/CreatePage.test.js]: No real reason.  I just glazed over it after setting up the mocks.  Side note, I miss RSpec's mock with argument constraints `allow(obj).to receive(:message).with('an argument').and_return(...)`.

Comment (bkemper) [reply src/pages/suppressions/tests/CreatePage.test.js]: Addressed... e97c1592

File: src/pages/suppressions/CreatePage.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (bkemper) [reply src/pages/suppressions/CreatePage.js]: Addressed... f3a8dbef

**Comment by bkemper** [comment]
@jasonrhodes This is ready for your final review.  Please let me know if you want to do a final walkthrough.

File: src/actions/suppressions.js [hunk 9]
```diff
@@ -116,9 +116,30 @@ export function deleteSuppression(suppression) {
...
+    // Convert deprecated type fields
+    if (!type && LIKE_TRUE.test(transactional)) { type = 'transactional'; }
+    if (!type && LIKE_TRUE.test(non_transactional)) { type = 'non_transactional'; }
+
+    // Format type value to provide a better user experience
+    if (type && LIKE_NON.test(type)) { type = 'non_transactional'; }
+    if (type && !LIKE_NON.test(type)) { type = 'transactional'; }
+
+    // Convert deprecated recipient fields
+    if (email) { recipient = email; }
```
Comment (jasonrhodes) [inline src/actions/suppressions.js]: `if (email && !recipient)` ?

File: src/actions/suppressions.js [hunk 10]
```diff
@@ -116,9 +116,30 @@ export function deleteSuppression(suppression) {
...
+
+    // Trim whitespace from recipient email
+    if (recipient) { recipient = _.trim(recipient); } // for FAD-5095
```
Comment (jasonrhodes) [inline src/actions/suppressions.js]: is `recipient` guaranteed at this point?

File: src/actions/tests/suppressions.test.js [hunk 11]
```diff
@@ -66,6 +66,32 @@ describe('Action Creator: Suppressions', () => {
...
+    it('with deprecated non_transactional and transactional field', () => {
+      const action = suppressions.createOrUpdateSuppressions([
+        { recipient: 'example@test.com', non_transactional: 'true', transactional: 'false' },
+        { recipient: 'example@test.com', non_transactional: 'false', transactional: 'true' }
+      ]);
+
+      expect(action).toMatchSnapshot();
+    });
+
+    it('with typo in type field', () => {
```
Comment (jasonrhodes) [inline src/actions/tests/suppressions.test.js]: maybe "with non-standard type fields, parses based on regex" or something? had trouble remembering what this was testing at first glance

File: src/actions/tests/suppressions.test.js [hunk 12]
```diff
@@ -66,6 +66,32 @@ describe('Action Creator: Suppressions', () => {
...
+    });
+
+    it('with deprecated non_transactional and transactional field', () => {
+      const action = suppressions.createOrUpdateSuppressions([
+        { recipient: 'example@test.com', non_transactional: 'true', transactional: 'false' },
+        { recipient: 'example@test.com', non_transactional: 'false', transactional: 'true' }
+      ]);
+
+      expect(action).toMatchSnapshot();
+    });
```
Comment (jasonrhodes) [inline src/actions/tests/suppressions.test.js]: Can we add a test where there is both `type` _and_ `transactional` `non_transactional` keys, to prove that type has precedence?

File: src/actions/suppressions.js (hunk 10 above)
Comment (bkemper) [reply src/actions/suppressions.js]: Yes.  I'll remove that condition.

File: src/actions/tests/suppressions.test.js [hunk 13]
```diff
@@ -66,6 +66,32 @@ describe('Action Creator: Suppressions', () => {
...
+
+    it('with typo in type field', () => {
+      const action = suppressions.createOrUpdateSuppressions([
+        { recipient: 'example@test.com', type: 'non_marketing' },
+        { recipient: 'example@test.com', type: 'marketing' }
+      ]);
+
+      expect(action).toMatchSnapshot();
+    });
```
Comment (jasonrhodes) [inline src/actions/tests/suppressions.test.js]: Add a test where there is both `recipient` and `email` keys to prove precedence?

File: src/actions/helpers/localFileParseRequest.js [hunk 14]
```diff
@@ -1,6 +1,18 @@
 import Papa from 'papaparse';
 
-export default ({ meta, type, parser = Papa }) => async(dispatch) => {
+export const hasData = ({ data }) => {
+  if (data.length === 0) {
+    return 'No data was found.';
+  }
+};
+
+export const hasField = (...expectedFields) => ({ meta: { fields }}) => {
```
Comment (jasonrhodes) [inline src/actions/helpers/localFileParseRequest.js]: Agree on renaming this, maybe `hasAtLeastOneField` lol idk

**Comment by jasonrhodes** [comment]
@bkemper a few general thoughts and questions:
* does papaparse do anything but CSV parsing? if not, should that action flow be named around CSV instead of more generic local file parse?
* for my own learning, what's the advantage of using a getter over a method?
* looks like you are calling meta.validate as if it is a single function but it's actually an array of functions, so that's blowing up

File: src/actions/helpers/localFileParseRequest.js [hunk 15]
```diff
@@ -23,10 +38,21 @@ export default ({ meta, type, parser = Papa }) => async(dispatch) => {
...
 
         reject(error);
-      } else {
-        dispatch({ type: types.SUCCESS, payload: { data }});
-        resolve(data);
+        return;
       }
+
+      // Custom validation of data
+      if (validate && (message = validate({ data, meta }))) {
```
Comment (jasonrhodes) [inline src/actions/helpers/localFileParseRequest.js]: Here's the bug ^

Should be able to fix by reducing over the validate array or something like that?

File: src/actions/tests/suppressions.test.js (hunk 11 above)
Comment (bkemper) [reply src/actions/tests/suppressions.test.js]: Addressed... 00eee30a

File: src/actions/suppressions.js (hunk 9 above)
Comment (bkemper) [reply src/actions/suppressions.js]: Addressed... 34f03d30

File: src/actions/suppressions.js (hunk 10 above)
Comment (bkemper) [reply src/actions/suppressions.js]: Addressed... a8de3ad2

File: src/actions/tests/suppressions.test.js (hunk 12 above)
Comment (bkemper) [reply src/actions/tests/suppressions.test.js]: Addressed... dde53c89

**Comment by bkemper** [comment]
> does papaparse do anything but CSV parsing? if not, should that action flow be named around CSV instead of more generic local file parse?

No, papaparse only handles CSV files.  Addressed... 10c87187

> for my own learning, what's the advantage of using a getter over a method?

Oops.  I forgot to switch it back after refactoring it.  Addressed... a2f787b2

> looks like you are calling meta.validate as if it is a single function but it's actually an array of functions, so that's blowing up

Yikes!  Addressed... 461a48c0

File: src/actions/helpers/csvFileParseRequest.js [hunk 16]
```diff
@@ -41,8 +41,10 @@ export default ({ meta: { file, validate }, type, parser = Papa }) => async(disp
         return;
       }
 
-      // Custom validation of data
-      if (validate && (message = validate({ data, meta }))) {
+      // Return the first validation error message
+      const message = _.find(validate, (v) => v({ data, meta }));
```
Comment (jasonrhodes) [inline src/actions/helpers/csvFileParseRequest.js]: oh good idea :clap:

File: src/actions/helpers/csvFileParseRequest.js [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+            details: errors
+          }
+        });
+
+        reject(error);
+        return;
+      }
+
+      // Return the first validation error message
+      const message = _.find(validate, (v) => v({ data, meta }));
```
Comment (jasonrhodes) [inline src/actions/helpers/csvFileParseRequest.js]: I thought this was an awesome idea until I realized it didn't quite work, doh. Haha.

See #282

Comment (bkemper) [reply src/actions/helpers/csvFileParseRequest.js]: Double yikes!  I should have seen this in the snapshots, `"message": [Function],`.

**Pull request merged**